components/TextFields/MDCTextInput.h

```diff
@@ -0,0 +1,159 @@
+   edited. The character count is below text.
+   */
+  MDCTextInputPresentationStyleFloatingPlaceholder,
+
+  /** The placeholder is laid out inline and the character count is also inline to the right. */
+  MDCTextInputPresentationStyleFullWidth,
+};
+
+/** Delegate for MDCTextInput size changes. */
+@protocol MDCTextInputLayoutDelegate <NSObject>
+
+@optional
+/**
+  Notifies the delegate that the textInput's content size changed, requiring the size provided.
+
+  @param textInput The text input for which the content size changed.
+  @param size      The size required by the text input to fit all of its content.
+ */
+- (void)textInput:(_Nonnull id<MDCTextInput>)textInput contentSizeChanged:(CGSize)size;
+
+@end
+
+/** Behavior exhibited by Material Design themed text inputs. */
+@protocol MDCTextInput <NSObject>
+
+/** A Boolean value indicating whether the text field is currently in edit mode. */
+@property(nonatomic, readonly, getter=isEditing) BOOL editing;
+
+/** The text to be displayed when no text has been entered. */
+@property(nonatomic, nullable, copy) NSString *placeholder;
+
+/** The text displayed in the text input. */
+@property(nonatomic, nullable, copy) NSString *text;
+
+/**
+ The text displayed under input rectangle, when in error state, that explains the nature of the
+ error.
+  The value of this property controls the state of the text input. When errorText != nil, the text
+  input is in an error state:
+   - The error text appears in a label below the input rectangle's underline with the errorColor as
+ text
+  color.
+   - The input rectangle's underline is colored with the errorColor.
+ */
+@property(nonatomic, nullable, copy, readonly) NSString *errorText;
+
+/**
+ The color used to denote error state in the underline and the errorText's label.
+ Default is red.
+ */
+@property(nonatomic, null_resettable, copy, readonly) UIColor *errorColor;
+
+/**
+ A localized string that represents the value of the errorText label. Use only when the
+ you need to override the default which is the errorText itself.
+ */
+@property(nonatomic, nullable, copy, readonly) NSString *errorAccessibilityText;
+
+/** The delegate for the text input. */
+@property(nonatomic, nullable, weak) id<MDCTextInputLayoutDelegate> textInputDelegate;
+
+/**
+  The presentation style of the text input.
+
+  Default is MDCTextInputPresentationStyleDefault.
+  */
+@property(nonatomic, assign) MDCTextInputPresentationStyle presentationStyle;
+
+/**
+  The tint applied to the caret, character count, floating placeholder and underline.
+
+  Default is black with Material Design hint text opacity.
+ */
+@property(nonatomic, null_resettable, strong) UIColor *tintColor;
+
+/**
+  The character limit for the text input.
+
+  Displays the character limit text taking into account the MDCTextInputPresentationStyle. A value
+  of 0 removes the limit text.
+
+  Default is 0.
+ */
+@property(nonatomic, assign) NSUInteger characterLimit;
```

Comment:
Has to be in MDCTextInput so it's in single line too: https://material.io/guidelines/components/text-fields.html#text-fields-character-counter